src/ch9329.cpp

```diff
@@ -0,0 +1,401 @@
+
+{
+  m_data = { 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+}
+
+ch9329::mouse_relative::mouse_relative()
+{
+  m_data = { 0x01, 0x00, 0x00, 0x00, 0x00 };
+}
+
+ch9329::keyboard_media::keyboard_media()
+{
+  m_data = { 0x02, 0x00, 0x00, 0x00 };
+}
+
+ch9329::keyboard_acpi::keyboard_acpi()
+{
+  m_data = { 0x01, 0x00 };
+}
+
+ch9329::keyboard_general::keyboard_general()
+{
+  m_data = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+}
+
+hal::byte get_size_byte(hal::byte p_command)
+{
+  switch (p_command) {
+    case cmd_send_kb_general_data:
+      return 0x08;
+    case cmd_send_kb_media_data:
+      return 0x04;
+    case cmd_send_ms_abs_data:
+      return 0x07;
+    case cmd_send_ms_rel_data:
+      return 0x05;
+  }
+  return 0x00;
+}
+
+void send_start_bytes(serial& p_serial, hal::byte p_command, uint8_t p_size = 0)
+{
+  std::array<hal::byte, header_frame_size> start_bytes = {
+    header_byte_1, header_byte_2, address_byte, p_command
+  };
+  if (p_size == 0) {
+    start_bytes[4] = get_size_byte(p_command);
+  } else {
+    start_bytes[4] = p_size;
+  }
+  hal::print(p_serial, start_bytes);
+}
+
+hal::byte calculate_sum(std::span<hal::byte> p_bytes, hal::byte p_command)
+{
+  std::uint8_t sum_byte = header_byte_1;
+  sum_byte += header_byte_2;
+  sum_byte += address_byte;
+  sum_byte += p_command;
+  sum_byte += get_size_byte(p_command);
+  for (hal::byte byte : p_bytes) {
+    sum_byte += byte;
+  }
+  return sum_byte;
+}
+
+void ch9329::send(ch9329::mouse_relative const& p_data)
+{
+  auto bytes = p_data.get_data();
+  send_start_bytes(*m_uart, cmd_send_ms_rel_data);
+  hal::print(*m_uart, bytes);
+  auto sum_byte = calculate_sum(bytes, cmd_send_ms_rel_data);
+  hal::print(*m_uart, std::to_array({ sum_byte }));
+}
+
+void ch9329::send(ch9329::mouse_absolute const& p_data)
+{
+  auto bytes = p_data.get_data();
+  send_start_bytes(*m_uart, cmd_send_ms_abs_data);
+  hal::print(*m_uart, bytes);
+  auto sum_byte = calculate_sum(bytes, cmd_send_ms_abs_data);
+  hal::print(*m_uart, std::to_array({ sum_byte }));
+}
+
+void ch9329::send(ch9329::keyboard_general const& p_data)
+{
+  auto bytes = p_data.get_data();
+  send_start_bytes(*m_uart, cmd_send_kb_general_data);
+  hal::print(*m_uart, bytes);
+  auto sum_byte = calculate_sum(bytes, cmd_send_kb_general_data);
+  hal::print(*m_uart, std::to_array({ sum_byte }));
+}
+
+void ch9329::send(keyboard_media const& p_data)
+{
+  auto bytes = p_data.get_data();
+  send_start_bytes(*m_uart, cmd_send_kb_media_data);
+  hal::print(*m_uart, bytes);
+  auto sum_byte = calculate_sum(bytes, cmd_send_kb_media_data);
+  hal::print(*m_uart, std::to_array({ sum_byte }));
+}
+
+void ch9329::send(keyboard_acpi const& p_data)
+{
+  auto bytes = p_data.get_data();
+  send_start_bytes(*m_uart, cmd_send_kb_media_data, 2);
+  hal::print(*m_uart, bytes);
+  auto sum_byte = calculate_sum(bytes, cmd_send_kb_media_data);
+  hal::print(*m_uart, std::to_array({ sum_byte }));
+}
```

Comment:
These are all pretty much identical besides the command byte. Maybe we can make another helper function that performs all of these operations for you. It could take a `p_data.get_data()` as a parameter as well.